src/main/java/org/ngafid/flights/CachedCSVWriter.java

```diff
@@ -0,0 +1,174 @@
+        }
+
+        if (!this.zipFile.canRead()) {
+            System.err.println("ERROR: do not have read access to this file!");
+            System.exit(1);
+        } else {
+            System.out.println("file is readable!");
+        }
+
+        try {
+            //this.zipArchive = new ZipFile(this.zipFile);
+            this.zipArchive = new ZipFile(archiveFilename);
+        } catch (IOException e) {
+            e.printStackTrace();
+            System.exit(1);
+        }
+
+        String filename = flight.getFilename();
+        Enumeration<? extends ZipEntry> entries = zipArchive.entries();
+        while (entries.hasMoreElements()) {
+            ZipEntry entry = entries.nextElement();
+
+            if (entry.getName().equals(filename)) {
+                this.entry = entry;
+            } 
+        } 
+    }
+
+    /**
+     * Gets the CSV file as primitive (binary) data
+     *
+     * @return a primitive array of bytes 
+     */
+    public byte [] toBinaryData() throws IOException {
+        InputStream inputStream = zipArchive.getInputStream(this.entry);
+        return IOUtils.toByteArray(inputStream);
+    }
+        
+    /**
+     * Accessor method for the {@link ZipEntry} associated with this flight
+     *
+     * @return an instance of {@link ZipEntry}
+     */
+    public ZipEntry getFlightEntry() {
+        return this.entry;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public String getFileContents() {
+        String fileOut = "";
+
+        try {
+            fileOut = new String(toBinaryData());
+            zipArchive.close();
+        } catch (IOException ie) {
+            ie.printStackTrace();
+        }
+
+        return fileOut;
+    }
+
+    /**
+     * Writes to a file and gets the zip archive first
+     *
+     * @return a String with the file contents
+     *
+     * @throws IOException if there is a problem with file i/o
+     */
+    public ZipEntry getZipEntry() throws IOException {
+        ZipFile zipArchive = new ZipFile(this.zipFile);
+        String filename = super.flight.getFilename();
+        Enumeration<? extends ZipEntry> entries = zipArchive.entries();
+
+        while (entries.hasMoreElements()) {
+            ZipEntry entry = entries.nextElement();
+
+            if (entry.getName().equals(filename)) {
+                zipArchive.close();
+                return entry;
+            } 
+        } 
+
+        zipArchive.close();
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void writeToFile() {
+        if (super.outputCSVFile.isPresent()) {
+            try {
+                FileOutputStream fos = new FileOutputStream(this.outputCSVFile.get());
+
+                fos.write(toBinaryData());
+                fos.close();
+            } catch (IOException ie) {
+                ie.printStackTrace();
+            }
+        } else {
+            // This should not happen!
+            return;
+        }
```

Comment:
Is there a specific reason to include the else block in both your Cached and Generated CSV Writers, if it just exits the function regardless?